L69-cpp-core-event-manager.md

```diff
@@ -0,0 +1,282 @@
+
+  // Returns the underlying TCP or UDP handle. Valid until Close() is called
+  // (i.e. not thread-safe with Close()).
+  virtual uv_handle_t fd() const = 0;
+
+  // Retruns the EventManager object where this descriptor is registered.
+  virtual LibuvEventManagerInterface* event_manager() const = 0;
+
+  // Executes all the registered notification closures, and closes the
+  // underlying TCP or UDP handle.
+  //
+  // Note that we cannot release the underlying TCP or UDP handle, since it is
+  // initialized with the libuv event loop of LibuvEventManager, and it is not
+  // thread-safe to access the handle externally.
+  virtual void Close() = 0;
+  virtual bool IsClosed() const = 0;
+
+  // Registers a closure \a upcall to be run immediately in a worker thread of
+  // the LibuvEventManager, after libuv reads from the socket. \a alloc_cb will
+  // be passed to libuv to allocate the read buffer. Note that this will start
+  // a read asynchronously.
+  //
+  // It is an error to call NotifyAfterRead() again before the previous closure
+  // completes.
+  //
+  // If there is more to read on the socket, re-arms a closure after upcall
+  // finishes.
+  virtuar void NotifyAfterRead(std::function<void(LibuvReadStatus)> upcall,
+                       uv_alloc_cb alloc_cb) = 0;
+
+  // Registers a closure \a upcall to be run immediately in a worker thread of
+  // the LibuvEventManager, after libuv writes to the socket. \a info contains
+  // the write message, and target address in cases of UDP. Note that this will
+  // start a write asynchronously.
+  //
+  // It is an error to call NotifyAfterWrite() again before the previous closure
+  // completes.
+  //
+  // If there is more to write, re-arms a closure after upcall finishes.
+  virtual void NotifyAfterWrite(std::function<void(int)> upcall, LibuvWriteInfo info) = 0;
+
+  // Accessor and mutator of arbitrary user-defined data. NOT thread-safe.
+  virtual void* data() = 0;
+  virtual void set_data(void* d) = 0;
+};
+
+class LibuvEventManagerInterface: public BaseEventManagerInterface {
+ public:
+  virtual ~LibuvEventManagerInterface();
+
+  virtual LibuvDescriptorInterface* RegisterFileDescriptor(uv_handle_t* fd) = 0;
+
+  virtual void DeleteDescriptor(LibuvDescriptorInterface* d) = 0;
+};
+```
+Registration of an EventManager implementation with Core
+User implementations of EventManager can be registered with gRPC Core by the following API:
+```cpp
+// Register a user-provided EventManager, which will be used globally in
+// gRPC. gRPC won't take the ownership of this EventManager; the
+//  EventManager must outlive gRPC.
+//
+// The function must be called before gRPC is initialized (i.e. grpc_init is first called).
+// It can only be called at most once; otherwise an exception will be thrown. If the function 
+// is not called at all, gRPC will use its internal default EventManager implementation.
+void grpc_register_global_event_manager(BaseEventManagerInterface *em);
```

Comment:
As the definition of `LibuvEventManagerInterface` and `EpollDescriptorInterface` showed, a working event manager requires not only implement the task managing logic, but also I/O handling. They are tightly coupled with the [grpc_endpoint](https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/endpoint.h) logic. So, there might be several solutions:

1. Let user provide grpc_endpoint implementation (vtable);
2. Alter the definition of event manager to include the logic currently in grpc_endpoint;
3. Instead of allow user-provided event manager, we say only the two implementation provided by us is supported.

Reply:
We'll require one implementation for each eventmanager implementation.

Reply:
@markdroth - does that not mean we'll have to expose grpc_endpoint interface too?

Reply:
> As the definition of `LibuvEventManagerInterface` and `EpollDescriptorInterface` showed, a working event manager requires not only implement the task managing logic, but also I/O handling. They are tightly coupled with the [grpc_endpoint](https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/endpoint.h) logic.

This is a good point.  I think we do need some way to register both the EventManager impl and the corresponding descriptor implementation.  We'll probably need to define some sort of common interface for descriptors to make this work.  But it seems like we'll also need some sort of factory interface to create new descriptors, and the factory is the thing that will probably need to be injected here along with the EM instance.

@guantaol, how are you making this work for the libuv EM?

> So, there might be several solutions:
> 
> 1. Let user provide grpc_endpoint implementation (vtable);

The issue here is descriptors, not endpoints.  Once we figure out how to have a base descriptor interface, we can figure out how to wire that into endpoints, but endpoints should not be exposed as part of this API.

Also, note that the current `grpc_endpoint` API is part of iomgr; ultimately, it will be removed along with the rest of the iomgr API.

> 2. Alter the definition of event manager to include the logic currently in grpc_endpoint;

We talked about this extensively, and it's not feasible due to the need to be compatible with the existing google3 EM API.

> 3. Instead of allow user-provided event manager, we say only the two implementation provided by us is supported.

Part of the goal of this API is to allow users to integrate gRPC into their own third-party event loops.  That won't work if users can't provide their own EM implementation.

Reply:
For libuv EM, it must work together with its associate endpoint implementation. Then for grpc runtime, it only needs to read from/write to the endpoints, but do not need to operate on the fd/socket directly.

Note that not only we need to consider the descriptor implementation with the event loop, but also take the polling model into consideration (NotifyAfterRead on Windows vs. NotifyWhenReadable on Unix). Thus, EM implementation is not platform-agnostic and we need an unified endpoint interface & implementation to wrap all the platform-specific details.

Reply:
That's a good point.  So how are you doing this for the libuv EM implementation?

For the API, it sounds like we have more work to do here.  We need to define a unified API that wraps both the EM and the endpoint implementation, so that users can override both of those things.  And once we do that, we might as well make that unified API the public API and not bother exposing the EM API at all.  This would certainly address the problem of leaking details of our existing internal EM API into this OSS API.

Guantao, I think this essentially reopens the discussion about inheritance vs. composition.  This unified interface is essentially exactly the composition approach that I was originally arguing for.  If we're going to need this anyway, then we're already going to need it to wrap the EventManager, in which case we are not actually getting the efficiency win that was the main benefit of the inheritance approach.  Remind me, were there other benefits to the inheritence approach?